File: backend/app/schemas/patient.py
from pydantic import BaseModel, EmailStr, Field
from typing import Optional
from datetime import datetime
from app.models.patient import PatientType, Gender


class PatientBase(BaseModel):
    first_name: str = Field(..., min_length=1, max_length=100)
    last_name: str = Field(..., min_length=1, max_length=100)
    date_of_birth: str
    gender: Gender
    phone: str = Field(..., min_length=1, max_length=20)
    email: Optional[EmailStr] = None
    address: Optional[str] = None
    patient_type: PatientType
    emergency_contact_name: Optional[str] = None
    emergency_contact_phone: Optional[str] = None
    blood_group: Optional[str] = None
    allergies: Optional[str] = None
    whatsapp_number: Optional[str] = None
    whatsapp_opt_in: Optional[str] = Field(default="no", pattern="^(yes|no)$")
    # Legacy field retained for compatibility with older frontend payloads
    emergency_contact: Optional[str] = None


class PatientCreate(PatientBase):
    tenant_id: Optional[str] = None


class PatientUpdate(BaseModel):
    first_name: Optional[str] = None
    last_name: Optional[str] = None
    date_of_birth: Optional[str] = None
    gender: Optional[Gender] = None
    phone: Optional[str] = None
    email: Optional[EmailStr] = None
    address: Optional[str] = None
    patient_type: Optional[PatientType] = None
    emergency_contact_name: Optional[str] = None
    emergency_contact_phone: Optional[str] = None
    blood_group: Optional[str] = None
    allergies: Optional[str] = None
    whatsapp_number: Optional[str] = None
    whatsapp_opt_in: Optional[str] = Field(default=None, pattern="^(yes|no)$")
    emergency_contact: Optional[str] = None


class PatientResponse(PatientBase):
    id: int
    mrn: str
    patient_id: str
    tenant_id: str
    is_active: bool
    created_at: datetime
    
    class Config:
        from_attributes = True

File: frontend/src/services/patient.ts
import { apiClient } from './api';

export interface Patient {
  id: number;
  mrn: string;
  patient_id: string;
  first_name: string;
  last_name: string;
  date_of_birth: string;
  gender: 'male' | 'female' | 'other';
  phone: string;
  email?: string;
  address?: string;
  patient_type: string;
  emergency_contact?: string; // legacy frontend compatibility
  emergency_contact_name?: string;
  emergency_contact_phone?: string;
  blood_group?: string;
  allergies?: string;
  whatsapp_number?: string;
  whatsapp_opt_in?: 'yes' | 'no';
  tenant_id: string;
  is_active: boolean;
  created_at: string;
}

export interface PatientCreate {
  first_name: string;
  last_name: string;
  date_of_birth: string;
  gender: 'male' | 'female' | 'other';
  phone: string;
  email?: string;
  address?: string;
  patient_type: string;
  emergency_contact?: string;
  emergency_contact_name?: string;
  emergency_contact_phone?: string;
  blood_group?: string;
  allergies?: string;
  whatsapp_number?: string;
  whatsapp_opt_in?: 'yes' | 'no';
  tenant_id: string;
}

export const patientService = {
  async getAll(): Promise<Patient[]> {
    return apiClient.get<Patient[]>('/api/v1/patients');
  },

  async getById(id: number): Promise<Patient> {
    return apiClient.get<Patient>(`/api/v1/patients/${id}`);
  },

  async create(data: PatientCreate): Promise<Patient> {
    return apiClient.post<Patient>('/api/v1/patients', data);
  },

  async update(id: number, data: Partial<PatientCreate>): Promise<Patient> {
    return apiClient.put<Patient>(`/api/v1/patients/${id}`, data);
  },

  async delete(id: number): Promise<void> {
    return apiClient.delete(`/api/v1/patients/${id}`);
  },

  async search(query: string): Promise<Patient[]> {
    return apiClient.get<Patient[]>('/api/v1/patients/search', { q: query });
  },
};
